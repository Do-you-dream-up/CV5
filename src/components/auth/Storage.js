import { cleanUrl, isDefined } from './helpers';

const store = localStorage;

const PKCE_KEY = 'pkce';
const TOKEN_KEY_ID = 'dydu-oauth-token-id';
const TOKEN_KEY_ACCESS = 'dydu-oauth-token-access';
const TOKEN_KEY_REFRESH = 'dydu-oauth-token-refresh';

export default class Storage {
  static savePkce(pkceData) {
    store.setItem(PKCE_KEY, JSON.stringify(pkceData));
  }

  static clearPkce() {
    store.removeItem(PKCE_KEY);
  }

  static loadPkce() {
    const pkce = store.getItem(PKCE_KEY);
    return isDefined(pkce) ? JSON.parse(pkce) : null;
  }

  static saveToken = (token) => {
    cleanUrl();
    Storage.clearPkce();
    localStorage.setItem(TOKEN_KEY_ID, token?.id_token);
    localStorage.setItem(TOKEN_KEY_ACCESS, token?.access_token);
    localStorage.setItem(TOKEN_KEY_REFRESH, token?.refresh_token);
  };

  static clearToken() {
    localStorage.removeItem(TOKEN_KEY_ID);
    localStorage.removeItem(TOKEN_KEY_ACCESS);
    localStorage.removeItem(TOKEN_KEY_REFRESH);
  }

  static loadToken() {
    return (
      isDefined(localStorage.getItem(TOKEN_KEY_ID)) && {
        id_token: isDefined(localStorage.getItem(TOKEN_KEY_ID)) && localStorage.getItem(TOKEN_KEY_ID),
        access_token: isDefined(localStorage.getItem(TOKEN_KEY_ACCESS)) && localStorage.getItem(TOKEN_KEY_ACCESS),
        refresh_token: isDefined(localStorage.getItem(TOKEN_KEY_REFRESH)) && localStorage.getItem(TOKEN_KEY_REFRESH),
      }
    );
  }

  static clearAll() {
    Storage.clearPkce();
    Storage.clearToken();
  }

  static containsPkce() {
    return Storage.loadPkce() !== null;
  }
}
