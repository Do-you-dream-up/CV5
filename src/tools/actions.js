const RE_ACTION =
  /javascript:\s*([\w.]+\s*\(\s*(?:'.*?'|".*?"|[^;]*)?(?:,\s*.*)*\))(?:\s*;+\s*([\w.]+\s*\(\s*(?:'.*?'|".*?"|[^;]*)?(?:,\s*.*)*\)))*/g;
const RE_ACTION_NAME = /^(\w+(?:.\w+)*)\s*\(\s*(.+)?\)/;
const RE_ACTION_PARAMETERS = /(?:^|,)\s*((('|").*?(\3))|([^,]+))/g;
const RE_ACTION_SANITIZE_STRING = /^('|")(.+)(\1)$/;

/**
 * Sanitize actions.
 */
export default (data) => {
  let actions = [];
  let response = [];
  let match = null;
  while ((match = RE_ACTION.exec(data))) {
    const [, ...matches] = match;
    actions = [...actions, ...matches.filter((it) => it)];
  }
  actions.map((it) => {
    let [, action, values] = it.match(RE_ACTION_NAME);
    let parameters = [];
    let match = null;
    while ((match = RE_ACTION_PARAMETERS.exec(values))) {
      parameters = [...parameters, match[0]];
    }
    parameters = parameters.map((it) => {
      const match = it.match(RE_ACTION_SANITIZE_STRING);
      return match ? match[2] : it;
    });
    response = [...response, { action, parameters }];
  });
  return response;
};
