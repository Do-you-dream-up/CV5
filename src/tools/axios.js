export const axiosConfigNoCache = {
  headers: {
    'Cache-Control': 'no-cache',
    Expires: '0',
  },
  params: {
    t: new Date().getTime(),
  },
};
