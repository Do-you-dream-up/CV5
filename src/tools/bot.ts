import { Local } from './storage';
import { getResourceWithoutCache } from './resources';
import { isDefined } from './helpers';

/**
 * - Wait for the bot ID and the API server then create default API based on the server ;
 * - Use BOT from local storage for the chatbox preview in Channels ;
 * - Protocol http is used when bliss is used in local with Channels.
 */

interface BotConfig {
  id: string;
  servers: string[];
  configId: string;
  getServer: (index: number) => string | null;
  getNextServer: (index: number, previousIndexUsed?: number) => string | null;
}

const isChannels = Local.isChannels.load();

export let BOT: BotConfig = {
  id: '',
  servers: [],
  configId: '',
  getServer: () => {
    return null;
  },
  getNextServer: () => {
    return null;
  },
};

export const getNextServerFromList = (servers: string[] | undefined, index: number, previousIndexUsed?: number) => {
  if (servers) {
    if (servers.length === 1) {
      return null;
    } else {
      let currentIndex: number = index;
      if (previousIndexUsed != undefined && previousIndexUsed !== index) {
        currentIndex = previousIndexUsed;
      }

      const nextIndex = currentIndex + 1;
      if (nextIndex >= servers.length) {
        return servers[0];
      }

      return servers[nextIndex];
    }
  }

  return null;
};

export const initBotInfoFromJsonOrChannels = async () => {
  const { data } = await getResourceWithoutCache('override/bot.json');

  let botData = {
    ...data,
    ...(isChannels && {
      id: Local.botId.load(),
      servers: Local.servers.load(),
    }),
  };

  const getServer = (index: number): string | null => {
    const servers: string[] = botData?.servers;
    return servers?.length > index && index >= 0 ? servers[index] : null;
  };

  const getNextServer = (index: number, previousIndexUsed?: number): string | null => {
    return getNextServerFromList(botData?.servers, index, previousIndexUsed);
  };

  botData = {
    ...botData,
    getServer: getServer,
    getNextServer: getNextServer,
  };

  BOT = Object.assign({}, botData);
  Local.botId.save(botData.id);
};
