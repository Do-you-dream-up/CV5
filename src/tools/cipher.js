import { Base64 } from 'js-base64';

/**
 * Deep-apply manipulator onto data.
 *
 * @param {*} data - Parameter description.
 * @param {function} manipulator - Function to transform with.
 * @returns {*} The transformed data.
 */
const transform = (data, manipulator) => {
  if (Array.isArray(data)) {
    for (let i = 0; i < data.length; i++) {
      data[i] = transform(data[i], manipulator);
    }
  } else if (typeof data === 'object') {
    for (let key in data) {
      data[key] = transform(data[key], manipulator);
    }
  } else if (typeof data === 'string') {
    return manipulator(data);
  }
  return data;
};

export const decode = (data) => transform(data, Base64.decode);
export const encode = (data) => transform(data, Base64.encode);
