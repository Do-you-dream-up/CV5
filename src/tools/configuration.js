import axios from 'axios';
import React from 'react';
import { ConfigurationContext } from '../contexts/ConfigurationContext';
import json from './configuration.json';
import { Local } from './storage';

/**
 * Helper class to find values in a JSON configuration file.
 */
export const configuration = new (class Configuration {
  constructor() {
    this.configuration = {};
  }

  /**
   * Fetch the configuration JSON and save its content.
   *
   * @param {string} [path] - Configuration file path.
   * @returns {Promise}
   */
  initialize = (
    path = `${process.env.PUBLIC_URL}override/configuration.json`,
  ) => {
    this.configuration = JSON.parse(JSON.stringify(json));

    return axios.get(path).then(({ data }) => {
      if (Local.get(Local.names.wizard)) {
        this.configuration = Local.get(Local.names.wizard);
        return this.sanitize(JSON.parse(JSON.stringify(this.configuration)));
      } else {
        return this.sanitize(JSON.parse(JSON.stringify(data)));
      }
    });
  };

  /**
   * Clean up and ensure sane style values. This typically will convert styles
   * into values JSS can understand.
   *
   * @param {Object} data - Configuration object to sanitize.
   * @returns {Object}
   */
  sanitize = (data) => {
    const forgeStyles = ({
      background,
      bottom,
      height,
      left,
      maxHeight,
      position,
      right,
      top,
      width,
    } = {}) => ({
      ...(background !== undefined && { background }),
      ...(bottom !== undefined && { bottom }),
      ...(height !== undefined && { height }),
      ...(left !== undefined && { left }),
      ...(maxHeight !== undefined && { maxHeight }),
      ...(position !== undefined && { position }),
      ...(right !== undefined && { right }),
      ...(top !== undefined && { top }),
      ...(width !== undefined && { width }),
    });
    Object.keys(data).forEach((key) => {
      if (data[key] instanceof Object) {
        data[key] = {
          ...data[key],
          ...(data[key].styles && { styles: forgeStyles(data[key].styles) }),
          ...(data[key].stylesMobile && {
            stylesMobile: forgeStyles(data[key].stylesMobile),
          }),
        };
      }
    });
    return data;
  };
})();

/**
 * High-order component to pass on configuration.
 */
export const withConfiguration = (Component) =>
  class InnerComponent extends React.PureComponent {
    static contextType = ConfigurationContext;
    render() {
      return (
        <Component configuration={this.context.configuration} {...this.props} />
      );
    }
  };
