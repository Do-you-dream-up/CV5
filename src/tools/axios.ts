import axios, { AxiosInstance, AxiosRequestConfig } from 'axios';
import { hasProperty, isDefined, isPositiveNumber, secondsToMs } from './helpers';
import { BOT, initBotInfoFromJsonOrChannels } from './bot';
import { getOidcEnableWithAuthStatus } from './oidc';
import Storage from '../components/auth/Storage';
import { decode } from './cipher';
import debounce from 'debounce-promise';
import dydu from './dydu';
import { Local } from './storage';

interface AxiosDyduConfig {
  server: string;
  timeout: number;
  axiosConf: any;
}

const sessionCurrentServerIndex = sessionStorage.getItem('dydu.server');
let currentServerIndex: number = sessionCurrentServerIndex ? Number.parseInt(sessionCurrentServerIndex) : 0;
let lastStatus = 'OK';
const minTimeoutForAnswer: number = secondsToMs(3);
let triesCounter = 0;
const maxTries = 3;

let configuration: any = null;
let lastResponse: any = null;

let callTokenRefresher: any = null;
let callOidcLogin: any = null;

const getAxiosInstanceWithDyduConfig = (config: AxiosDyduConfig) => {
  if (!isDefined(config?.axiosConf)) config.axiosConf = {};

  const instance = axios.create({
    baseURL: config?.server,
    timeout: isPositiveNumber(config?.timeout) ? config.timeout : axios.defaults.timeout,
    ...config.axiosConf,
  });

  // when request is sent
  instance.interceptors.request.use(
    (config: AxiosRequestConfig) => {
      if (getOidcEnableWithAuthStatus()) {
        const tokenInfo = Storage.loadToken();
        const idToken: string | null | undefined = isDefined(tokenInfo?.id_token) ? tokenInfo.id_token : undefined;
        const accessToken: string | null | undefined = isDefined(tokenInfo?.access_token)
          ? tokenInfo.access_token
          : undefined;

        if (!config.headers) {
          config.headers = {};
        }

        if (idToken) {
          config.headers['Authorization'] = `Bearer ${idToken}`;
        }

        if (accessToken) {
          config.headers['OIDCAccessToken'] = `${accessToken}`;
        }
      }
      return config;
    },
    (error) => {
      return Promise.reject(error);
    },
  );

  // when response code in range of 2xx
  const onSuccess = (response: any) => {
    return response;
  };

  instance.interceptors.response.use(onSuccess);

  return instance;
};

const handleSwitchToBackup = (currentServerPreviouslyUsed?: number) => {
  if (lastStatus === 'Error') {
    const backupServer: string | null = BOT.getNextServer(currentServerIndex, currentServerPreviouslyUsed);
    console.log('Status error, switching to server=' + backupServer);

    if (backupServer) {
      currentServerIndex++;
      if (currentServerIndex >= BOT.servers.length) {
        currentServerIndex = 0;
        triesCounter++;
        console.log('TriesCounter=' + triesCounter + ', waiting 5000ms');
        (async () => wait(5000))();
      }
      SERVLET_API.defaults.baseURL = buildServletApiUrl(backupServer);
      SERVLET.defaults.baseURL = buildServletUrl(backupServer);
    } else {
      triesCounter++;
      (async () => wait(5000))();
    }
  }
};

const wait = async (ms: number) => {
  await new Promise((resolve) => setTimeout(resolve, ms));
};

const handleSetApiTimeout = (ms: number) => {
  let timeout = minTimeoutForAnswer;
  if (ms) {
    timeout = ms;
  }

  if (SERVLET_API?.defaults) {
    SERVLET_API.defaults.timeout = timeout;
  }

  if (SERVLET?.defaults) {
    SERVLET.defaults.timeout = timeout;
  }
};

const handleAxiosResponse = (data: any) => {
  if (!data) {
    data = {};
  }

  data && configuration?.saml?.enable && samlRenewOrReject(data);

  if (hasProperty(data, 'values')) {
    data.values = decode(data.values);
    if (data.values.contextId) {
      dydu.setContextId(data.values.contextId);
    }
    return data.values;
  }

  return data;
};

const samlRenewOrReject = ({ type, values }) => {
  switch (type) {
    case 'SAML_redirection':
      redirectAndRenewAuth(values);
      break;
    default:
      return renewAuth(values?.auth);
  }
};

const redirectAndRenewAuth = (values: any) => {
  const relayState = encodeURI(window.location.href);
  // const relayState = JSON.stringify({ redirection: encodeURI(window.location.href), bot: BOT.id });
  try {
    renewAuth(atob(values?.auth));
    window.location.href = `${atob(values?.redirection_url)}&RelayState=${relayState}`;
  } catch {
    renewAuth(values?.auth);
    window.location.href = `${values?.redirection_url}&RelayState=${relayState}`;
  }
};

const renewAuth = (auth: string) => {
  if (auth) {
    try {
      Local.saml.save(atob(auth));
    } catch {
      Local.saml.save(auth);
    }
  }
};

const handleAxiosError = (
  error: any,
  verb: any,
  path: string,
  data: any,
  timeout: number,
  currentServerIndexForThisRequest: number,
) => {
  if (triesCounter >= maxTries) {
    throw 'API Unreachable';
  }

  /**
   * NO 401 ERROR
   */
  if (error?.response?.status !== 401) {
    lastStatus = 'Error';
  }

  /**
   * IF 401
   */
  if (getOidcEnableWithAuthStatus()) {
    return new Promise((resolve) =>
      handleTokenRefresh(() => {
        setTimeout(() => {
          return resolve(emit(verb, path, data, timeout, currentServerIndexForThisRequest));
        }, minTimeoutForAnswer);
      }),
    );
  }

  // Retry API Call
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(emit(verb, path, data, timeout, currentServerIndexForThisRequest));
    }, minTimeoutForAnswer);
  });
};

const handleTokenRefresh = (retry: any) => {
  if (configuration?.oidc?.enable) {
    const tokenInfo = Storage.loadToken();
    const refreshToken = isDefined(tokenInfo?.refresh_token) ? tokenInfo.refresh_token : undefined;

    if (refreshToken && refreshToken?.length > 0 && triesCounter < 2 && Storage.loadPkceCodeVerifier()) {
      console.log('Refreshing token...');
      callTokenRefresher().then(() => {
        return retry();
      });
    } else {
      console.log('No refresh token found, redirecting to login page...');
      Storage.clearToken();
      callOidcLogin();
      return Promise.resolve();
    }
  }
};

const formatConsoleError = (error: any) => {
  console.group('An error occurred while sending the request to Dydu. Please send us this group of logs :');
  console.log('Date : ', new Date());
  console.log('Error : ', error.message);
  console.log('Status : ', error.status || error?.response?.status);
  console.log('Request : [%s] %s%s', error.config.method.toUpperCase(), error.config.baseURL, error.config.url);
  console.log('Headers :', JSON.stringify(error.config.headers));
  console.log(
    'Data :\n%s',
    error.config.data
      ? error.config.data
          .trim()
          .split('&')
          .map((s: string) => decodeURIComponent(s))
          .join('\n')
      : 'no data',
  );
  console.groupEnd();
};

/**
 * Request against the provided path with the specified data. When
 * the response contains values, decode it and refresh the context ID.
 * if the request fail several times the request will be failover to back-up server.
 *
 * @param {function} verb - A verb method to request with.
 * @param {string} path - Path to send the request to.
 * @param {Object} data - Data to send.
 * @param {number} timeout - timeout value to use.
 * @param {boolean} ignoreSwitch - set tu true, the server switch will not occur.
 * @param {number} currentServerPreviouslyUsed - server index used with the last request. Used to switch server
 * @returns {Promise}
 */
const emit = debounce(
  (verb: any, path: string, data: any, timeout: number, currentServerPreviouslyUsed?: number) => {
    handleSwitchToBackup(currentServerPreviouslyUsed);
    handleSetApiTimeout(timeout);

    const pathWithoutTrailingSlash = path.replace(/\/$/, '');

    const currentServerIndexForThisRequest = currentServerIndex;

    try {
      return verb(pathWithoutTrailingSlash, data)
        .then((response: any) => {
          lastResponse = response;
          lastStatus = 'OK';
          triesCounter = 0;
          sessionStorage.setItem('dydu.server', currentServerIndex.toString());
          return response;
        })
        .then(({ data = {} }) => {
          return handleAxiosResponse(data);
        })
        .catch((error: any) => {
          formatConsoleError(error);
          return handleAxiosError(
            error,
            verb,
            pathWithoutTrailingSlash,
            data,
            timeout,
            currentServerIndexForThisRequest,
          );
        });
    } catch (e) {
      console.error('while executing |emit()|', e);
    }
  },
  100,
  { leading: true },
);

const getServerFromIndex = (index: number): string | null => {
  return BOT !== null ? BOT.getServer(index) : null;
};

const isLocalServer = (server: string | null): boolean => {
  return (
    server !== null &&
    (server.includes('dev.dydu.local') || server.includes('localhost') || server.endsWith('.loca.lt'))
  );
};

const buildServletUrl = (server?: string | null): string => {
  if (!server) {
    server = getServerFromIndex(currentServerIndex);
  }

  const servletPath = isLocalServer(server) ? '' : 'servlet/';
  return `${server}/${servletPath}`;
};

const buildServletApiUrl = (server?: string | null): string => {
  return buildServletUrl(server) + 'api/';
};

let SERVLET_API: AxiosInstance, SERVLET: AxiosInstance;

initBotInfoFromJsonOrChannels().then(() => {
  SERVLET_API = getAxiosInstanceWithDyduConfig({
    server: buildServletApiUrl(),
    timeout: minTimeoutForAnswer,
    axiosConf: {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8',
      },
    },
  });

  /**
   * Used for old Servlet API
   */
  SERVLET = getAxiosInstanceWithDyduConfig({
    server: buildServletUrl(),
    timeout: minTimeoutForAnswer,
    axiosConf: {
      headers: {
        Accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8',
        'Content-Type': 'multipart/form-data',
      },
    },
  });
});

const getLastResponse = (): any => lastResponse;
const setLastResponse = (newLastResponse: any) => (lastResponse = newLastResponse);

const setConfiguration = (newConfiguration: any) => (configuration = newConfiguration);

const setCallTokenRefresher = (newCallTokenRefresher: any) => {
  callTokenRefresher = newCallTokenRefresher;
};

const setCallOidcLogin = (newCallOidcLogin: any) => {
  callOidcLogin = newCallOidcLogin;
};

const emitLivechatError = () => {
  currentServerIndex++;
};

export {
  getLastResponse,
  setLastResponse,
  setConfiguration,
  setCallTokenRefresher,
  setCallOidcLogin,
  getAxiosInstanceWithDyduConfig,
  emit,
  emitLivechatError,
  buildServletUrl,
  SERVLET_API,
  SERVLET,
  currentServerIndex,
};
