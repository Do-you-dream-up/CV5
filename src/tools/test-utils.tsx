import { JssProvider, ThemeProvider } from 'react-jss';
import React, { ReactElement } from 'react';
import { RenderOptions, render } from '@testing-library/react';

import { ConfigurationProvider } from '../contexts/ConfigurationContext';
import { EventsProvider } from '../contexts/EventsContext';
import ViewModeProvider from '../contexts/ViewModeProvider';
import configuration from '../../public/override/configuration.json';
import theme from '../../public/override/theme.json';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <JssProvider>
      <ThemeProvider theme={theme}>
        <ConfigurationProvider configuration={configuration}>
          <ViewModeProvider>
            <EventsProvider>{children}</EventsProvider>
          </ViewModeProvider>
        </ConfigurationProvider>
      </ThemeProvider>
    </JssProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
