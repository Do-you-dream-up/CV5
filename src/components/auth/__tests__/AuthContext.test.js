/* eslint-disable react/prop-types */
import Storage from '../Storage';
import axios from 'axios';
import { render } from '../../../tools/test-utils';

const jwtToken =
  'eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJ0c1BTTUFsa1JxaGs2OEpxZEhaTWxiYWRfMV9OYjVMZ05PVUVrVm9ZS29RIn0.eyJleHAiOjE2NjMzMTI4MDgsImlhdCI6MTY2MzMxMjIwOCwianRpIjoiN2JkYzU4NTEtM2Y0OC00ZWU0LTk4NjAtOTYyZjk3MGQwNWZiIiwiaXNzIjoiaHR0cHM6Ly9rZXljbG9hay5zZWN1cml0eS5keWR1LXByaXYuY29tL2F1dGgvcmVhbG1zL2JhY2tfdGVzdCIsImF1ZCI6WyJyZWFsbS1tYW5hZ2VtZW50IiwiYWNjb3VudCJdLCJzdWIiOiIxNzBmNjA4Yy1hNmE0LTQ2NzctYWNmMy0wZDRjZDdjMGNkYjgiLCJ0eXAiOiJCZWFyZXIiLCJhenAiOiJhbGV4LWF1dGgiLCJzZXNzaW9uX3N0YXRlIjoiMjk4Y2JhN2QtMjNhMi00OTA0LTgxNjQtZjM1Y2I4NjM1ZjY5IiwiYWNyIjoiMSIsImFsbG93ZWQtb3JpZ2lucyI6WyJodHRwczovL2tleWNsb2FrLnNlY3VyaXR5LmR5ZHUtcHJpdi5jb20iXSwicmVhbG1fYWNjZXNzIjp7InJvbGVzIjpbIm9mZmxpbmVfYWNjZXNzIiwiZGVmYXVsdC1yb2xlcy1iYWNrX3Rlc3QiLCJ1bWFfYXV0aG9yaXphdGlvbiJdfSwicmVzb3VyY2VfYWNjZXNzIjp7InJlYWxtLW1hbmFnZW1lbnQiOnsicm9sZXMiOlsidmlldy1pZGVudGl0eS1wcm92aWRlcnMiLCJ2aWV3LXJlYWxtIiwibWFuYWdlLWlkZW50aXR5LXByb3ZpZGVycyIsImltcGVyc29uYXRpb24iLCJyZWFsbS1hZG1pbiIsImNyZWF0ZS1jbGllbnQiLCJtYW5hZ2UtdXNlcnMiLCJxdWVyeS1yZWFsbXMiLCJ2aWV3LWF1dGhvcml6YXRpb24iLCJxdWVyeS1jbGllbnRzIiwicXVlcnktdXNlcnMiLCJtYW5hZ2UtZXZlbnRzIiwibWFuYWdlLXJlYWxtIiwidmlldy1ldmVudHMiLCJ2aWV3LXVzZXJzIiwidmlldy1jbGllbnRzIiwibWFuYWdlLWF1dGhvcml6YXRpb24iLCJtYW5hZ2UtY2xpZW50cyIsInF1ZXJ5LWdyb3VwcyJdfSwiYWNjb3VudCI6eyJyb2xlcyI6WyJtYW5hZ2UtYWNjb3VudCIsIm1hbmFnZS1hY2NvdW50LWxpbmtzIiwidmlldy1wcm9maWxlIl19fSwic2NvcGUiOiJlbWFpbCBwcm9maWxlIiwic2lkIjoiMjk4Y2JhN2QtMjNhMi00OTA0LTgxNjQtZjM1Y2I4NjM1ZjY5IiwiZW1haWxfdmVyaWZpZWQiOnRydWUsIm5hbWUiOiJBbGV4YW5kcmUgTW9uZ2UiLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJhbW9uZ2UiLCJnaXZlbl9uYW1lIjoiQWxleGFuZHJlIiwiZmFtaWx5X25hbWUiOiJNb25nZSIsImVtYWlsIjoiYW1vbmdlQGR5ZHUuYWkifQ.FbMHD9c0v9_1rVkC5tOYBBwbOxHKOFbgGPM-xytQfc0JXNtgY9UxjyfJawglNcsC9kdIYmqDlSuI9B1ovlYav0-qiX7BRgjuAi18OzenJRPBgBH6yjLlMTdzyufZyDh0w7Xqa5RVYY_yrIw2HDBw4Cfz5agiGdJQvQrU4WYXqJNMvAWMcX1JQqEyj3xxoXNYc_Az03KVpsii0Z1hCP2N35N5yUn4x1QOZn8lfgHKvK_8l_r2U4JMeFyJ1bxZSNsWZJtlqIx5sevXHtAguLGH1klGaax33BBrYJmOBS1CQ-LLlmWNy9Bw8D84eWYUEoBG3dH_Y48fZUT8Yzt_Du-1Iw';

const oidcConfig = {
  enable: true,
  clientId: 'l5LUWFkMo18GaWJZvzXHde1B0Ehz236tBSdphOvrWN8hn88qbOEVpgvZU0kW2Hnu',
  clientSecret: '5CowOGbBLi5GOnIlwF6cpSJ5HLdAQwVEBwxvnbyeMeit1XhKsQetH9r7cU6hTzeO',
  pkceActive: true,
  pkceMode: 'S256',
  discoveryUrl: 'https://yawks.net/nextcloud/index.php/apps/oidc/openid-configuration',
  scopes: ['openid', 'email', 'profile'],
  withAuth: true,
};

jest.mock('../Storage', () => {
  let token = null;

  return {
    __esModule: true,
    default: {
      loadToken: jest.fn(() => token),
      clearToken: jest.fn(),
      saveUrls: jest.fn(),
      loadUrls: jest.fn(),
      clearUrls: jest.fn(),
      clearUserInfo: jest.fn(),
      saveToken: jest.fn((t) => {
        token = t;
      }),
      saveUserInfo: jest.fn((t) => {
        token = t;
      }),
      clearPkce: jest.fn(),
      loadPkce: jest.fn(() => token),
      savePkce: jest.fn((t) => {
        token = t;
      }),
      containsPkce: jest.fn(),
    },
  };
});

jest.mock('../../../tools/dydu', () => {
  let authorize = null;
  let botLanguages = ['en', 'fr'];

  return {
    __esModule: true,
    default: {
      setOidcLogin: jest.fn(() => authorize),
      setServerStatusCheck: jest.fn(),
      setTokenRefresher: jest.fn(),
      setSpaceToDefault: jest.fn(),
      fetchUrlConfig: jest.fn(),
      setLocale: jest.fn(),
      getBotLanguages: jest.fn(
        () =>
          new Promise((resolve) => {
            resolve(botLanguages);
          }),
      ),
    },
  };
});

jest.mock('axios');

describe('AuthProvider', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    Storage.clearUserInfo();
    Storage.clearUrls();
  });

  afterAll(() => {
    Storage.clearToken();
    Storage.clearPkce();
    Storage.clearUrls();
  });

  it('should save urls to storage and set url config', async () => {
    const data = {
      authorization_endpoint: 'http://example.com/authorize',
      token_endpoint: 'http://example.com/token',
      userinfo_endpoint: 'http://example.com/userinfo',
    };

    axios.get.mockResolvedValueOnce({ data });

    const urls = {
      authUrl: data.authorization_endpoint,
      tokenUrl: data.token_endpoint,
      userinfoUrl: data.userinfo_endpoint,
    };

    Storage.clearUserInfo();
    Storage.clearUrls();
    Storage.saveUrls(urls);

    expect(Storage.clearUserInfo).toHaveBeenCalled();
    expect(Storage.clearUrls).toHaveBeenCalled();
    expect(Storage.saveUrls).toHaveBeenCalledWith(urls);
  });

  test('renders children', () => {
    const { getByText } = render(<div>Test Children</div>);
    expect(getByText('Test Children')).toBeDefined();
  });

  test('renders children when enable is false', () => {
    const { getByText } = render(<div>Test Children</div>);
    expect(getByText('Test Children')).toBeDefined();
  });

  test('renders children when enable=true and user is logged in', () => {
    const mockToken = {
      access_token: jwtToken,
      refresh_token: jwtToken,
    };

    const mockUserInfo = { email: 'mgauchedaumet@dydu.ai' };

    Storage.saveToken(mockToken);
    // Storage.saveUserInfo(mockUserInfo);

    const { getByText } = render(<div>Protected Content</div>, {
      configuration: {
        oidc: oidcConfig,
      },
      auth: { clientId: 'test-client-id' },
    });
    expect(getByText('Protected Content')).toBeDefined();
  });

  test('renders children when enable=true and user is logged in but with false token', () => {
    const mockToken = {
      access_token: 'pasbon',
      refresh_token: 'pasbon',
    };

    Storage.saveToken(mockToken);

    const { getByText } = render(<div>Protected Content</div>, {
      configuration: {
        oidc: oidcConfig,
      },
      auth: { clientId: 'test-client-id' },
    });
    expect(getByText('Protected Content')).toBeDefined();
  });

  test('renders children when isLoadedFromChannels', () => {
    window.dyduReferer = true;
    const { getByText } = render(<div>Protected Content</div>);
    expect(getByText('Protected Content')).toBeDefined();
  });
});
