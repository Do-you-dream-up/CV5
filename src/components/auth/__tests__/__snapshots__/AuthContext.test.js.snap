// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthProvider renders children 1`] = `
<div>
  Test
</div>
`;
